core/test/forest/CausalForestTest.cpp

```diff
@@ -0,0 +1,78 @@
+/*-------------------------------------------------------------------------------
+  This file is part of generalized random forest (grf).
+
+  grf is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  grf is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with grf. If not, see <http://www.gnu.org/licenses/>.
+ #-------------------------------------------------------------------------------*/
+
+#include "commons/utility.h"
+#include "forest/ForestPredictor.h"
+#include "forest/ForestPredictors.h"
+#include "forest/ForestTrainer.h"
+#include "forest/ForestTrainers.h"
+#include "utilities/ForestTestUtilities.h"
+
+#include "catch.hpp"
+
+
+TEST_CASE("causal forests are invariant to rescaling of the sample weights", "[causal, forest]") {
+  // Run the original forest.
+  // we'll overwrite a covariate in the original data with sample weights so we needn't resize the data.
+  uint weight_index = 9;
+  uint outcome_index = 10;
+  uint treatment_index = 11;
+  Data* data = load_data("test/forest/resources/causal_data.csv");
+  data->set_weight_index(weight_index);
+  data->set_outcome_index(outcome_index);
+  data->set_treatment_index(treatment_index);
+  data->set_instrument_index(treatment_index);
+
+  bool error;
+  for(size_t r = 0; r < data->get_num_rows(); r++) {
+    double weight = 1.0 / (1.0 + exp(- data->get(1, r)));
```

Comment:
We're indexing by `(column, row)`, but the method signature is `double get(size_t row, size_t col)`, so this is going to produce an invalid read.

Reply:
I hadn't noticed, thanks. Isn't it odd that the order for get and set are reversed?

```
double DefaultData::get(size_t row, size_t col) const {
  return data[col * num_rows + row];
}
void DefaultData::set(size_t col, size_t row, double value, bool& error) {
  data[col * num_rows + row] = value;
}
```

Reply:
I flipped the arguments for the call to get in the test. Good to merge?